brainiak/matnormal/covs.py

```diff
@@ -3,438 +3,475 @@
         """
-        return tf.diag(tf.ones([self.size], dtype=tf.float64))
+        return tf.eye(self.size, dtype=tf.float64)
+
+    @property
+    def _cov(self):
+        """Expose the covariance explicitly (mostly for testing /
+        visualization, materializing large covariances may be intractable)
+        """
+        return tf.eye(self.size, dtype=tf.float64)
 
 
 class CovAR1(CovBase):
-    """AR1 covariance
+    """AR(1) covariance parameterized by autoregressive parameter rho
+    and new noise sigma.
+
+    Parameters
+    ----------
+    size: int
+        size of covariance matrix
+    rho: float or None
+        initial value of autoregressive parameter (if None, initialize
+        randomly)
+    sigma: float or None
+        initial value of new noise parameter (if None, initialize randomly)
+
     """
+
     def __init__(self, size, rho=None, sigma=None, scan_onsets=None):
 
+        super(CovAR1, self).__init__(size)
+
         # Similar to BRSA trick I think
         if scan_onsets is None:
             self.run_sizes = [size]
-            self.offdiag_template = tf.constant(scipy.linalg.toeplitz(np.r_[0,
-                                                1, np.zeros(size-2)]),
-                                                dtype=tf.float64)
-            self.diag_template = tf.constant(np.diag(np.r_[0,
-                                                           np.ones(size-2),
-                                                           0]))
+            self.offdiag_template = tf.constant(
+                scipy.linalg.toeplitz(np.r_[0, 1, np.zeros(size - 2)]),
+                dtype=tf.float64
+            )
+            self.diag_template = tf.constant(np.diag(
+                  np.r_[0, np.ones(size - 2), 0]))
         else:
             self.run_sizes = np.ediff1d(np.r_[scan_onsets, size])
-            sub_offdiags = [scipy.linalg.toeplitz(np.r_[0, 1, np.zeros(r-2)])
-                            for r in self.run_sizes]
-            self.offdiag_template = tf.constant(scipy.sparse.
-                                                block_diag(sub_offdiags)
-                                                .toarray())
-            subdiags = [np.diag(np.r_[0, np.ones(r-2), 0])
+            sub_offdiags = [
+                scipy.linalg.toeplitz(np.r_[0, 1, np.zeros(r - 2)])
+                for r in self.run_sizes
+            ]
+            self.offdiag_template = tf.constant(
+                scipy.sparse.block_diag(sub_offdiags).toarray()
+            )
+            subdiags = [np.diag(np.r_[0, np.ones(r - 2), 0])
                         for r in self.run_sizes]
-            self.diag_template = tf.constant(scipy.sparse.
-                                             block_diag(subdiags)
-                                             .toarray())
+            self.diag_template = tf.constant(
+                scipy.sparse.block_diag(subdiags).toarray()
+            )
```

Comment:
Side note: I don't know how efficiently tensorflow handles these block diagonal matrices. If it is not very efficient, I wonder if (potentially in a future update), you would consider just separately computing the log likelihood for each run. Because basically there would be no temporal correlation between runs, and I think the total log likelihood is just the sum of each.

Reply:
Good point. Shall I write up an issue for this after merging? Prefer not to incorporate it into this PR, lots of things here already.